src/openms/source/CONCEPT/QCChargeDistribution.cpp

```diff
@@ -0,0 +1,156 @@
+{
+
+}
+
+//Constructor
+QCChargeDistribution::QCChargeDistribution(std::vector<std::pair<OpenMS::String,OpenMS::FeatureMap>> files):
+  feat_map_(files)
+  {
+      
+  };
+  
+//Main method to write mztab peptide section data needed for charge distribution plot (PTXQC)
+int QCChargeDistribution::ChargeDistribution(MzTab& mztab) const
+{
+  vector<FeatureMap> maps;
+  MzTabPeptideSectionRows rows;
+  MzTabPeptideSectionRows mztabRows = mztab.getPeptideSectionRows();
+  int pepIDCount = 0;
+
+  for(vector<pair<String,FeatureMap>>::const_iterator it = feat_map_.begin();it!=feat_map_.end();++it)
+  {
+    maps.push_back (it->second);
+  }
+   
+  for (unsigned long m = 0; m < maps.size(); m++)
+  {     
+  
+    String rfile;
+    
+    if (maps[m].metaValueExists("spectra_data")) 
+    {		
+      StringList rfiles = maps[m].getMetaValue("spectra_data");	
+      rfile = rfiles[0];
+    } 
+    
+    for (vector<Feature>::const_iterator f_it = maps[m].begin(); f_it!=maps[m].end();f_it++)
+    {
+
+      vector<PeptideIdentification> pep_id = f_it->getPeptideIdentifications();	
+ 		  
+      if (pep_id.empty()) 
+      {
+        //Empty lines peptide_data_ for features with retention times	
+      }
+ 				
+      else 
+      {
+ 	    for (vector<PeptideIdentification>::iterator p_it = pep_id.begin(); p_it!=pep_id.end(); p_it++) 
+ 	    {
+ 	      pepIDCount++;
+ 	      MzTabPeptideSectionRow row;
+ 	      MzTabString PepSeq;
+ 	      MzTabInteger charge;
+ 		  
+ 		  //Set sequence and charge	
+          vector<PeptideHit> hits = p_it->getHits(); 
+          PeptideHit hit = hits[0];
+          AASequence seq = hit.getSequence();
+          int ch = hit.getCharge();
+          PepSeq.set(seq.toString());
+          charge.set(ch);
+          row.charge = charge;
+          row.sequence = PepSeq;
+          
+    	  //Set optional columns: original RT and source file
+          vector<MzTabOptionalColumnEntry> v;                
+          MzTabString name = MzTabString(rfile);
+          MzTabOptionalColumnEntry sraw = make_pair("raw_source_file",name);
+          v.push_back (sraw);
+          row.opt_ = v;
+                
+          rows.push_back(row);
+    	  		
+         }   
+ 	  }
+ 	    
+ 	}			
+    		
+  }
+  
+  //If peptide section was written from features before: append columns
+  if (pepIDCount==mztabRows.size()) 
+  {
+    for (unsigned i = 0; i < mztabRows.size(); i++)
+    {
+      MzTabPeptideSectionRow mz_r = mztabRows[i];
+      MzTabPeptideSectionRow r = rows[i];
+      MzTabInteger ch = r.charge;
+      vector<MzTabOptionalColumnEntry> v = r.opt_;
+      mz_r.charge = ch;
+      vector<MzTabOptionalColumnEntry> mz_v = mz_r.opt_;
+      for(unsigned j = 0; j < v.size(); j++)
+      {
+        mz_v.push_back(v[j]);
+      }
+      mz_r.opt_ = mz_v;
+      mztabRows[i] = mz_r;
+    }
+    mztab.setPeptideSectionRows(mztabRows);
+  }
+  //Else: append rows
+  else
+  {
+    for (unsigned i = 0; i < rows.size(); i++)
```

Comment:
this is highly unstable.
Rows should be added "in order" not at the bottom. This could be done via UniqueIDInterface, for example.